### PR TITLE
Multigroup Per-Thread Cache Conversion

### LINKED ISSUE #2590: Multigroup MC Gives Wrong Answer in Event-Based Mode
## Bug Description

OpenMC gives incorrect answers (e.g., eigenvalue, tallies) when running in event-based mode and when using multigroup cross section data that features materials with multiple temperatures and/or multiple angles (i.e., non-isotropic). OpenMC works correctly in event-based mode with isotropic + isothermal MGXS materials.

The bug is caused by the per-thread MGXS cache. Thread-owned particle state does not transfer between events, so does not work in event-based mode. The code still runs correctly if isotropic and isothermal data is used because in these cases, the cache will always store `temperature` and `angle` indices of `0`, such that it doesn't matter when/by who the cache is accessed or written to/read from.

## Steps to Reproduce

The issue can be seen by changing the `openmc/examples/pincell_multigroup` example to include another temperature level for one of the materials (see code below). I chose to split the fueled region of the pincell into an inner cylinder at higher temperature and an outer ring at lower temperature. I didn't actually generate new high temperature XS data, I simply copied the original MGXS data for the fueled region but permuted the Chi distribution. This accomplished the goal of having a single material with multiple temperatures each with different XS data, such that after a few batches any errors in OpenMC will cause divergence.

I then ran the code in a variety of configurations with 5 inactive batches, 5 active batches, 10k particles/batch, gcc 12.2, on a dual-socket 8180M system with 56c/112t, as seen below:

Parallelism | Branch | Mode | k-eff
-- | -- | -- | --
Serial | Thread Cache (Baseline) | history | 1.32339
Serial | Thread Cache (Baseline) | event | 1.32117
MPI 112 ranks | Thread Cache (Baseline) | event | 1.31853
OpenMP 112 threads | Thread Cache (Baseline) | event | 1.32285

All the different cases produce different results. The correct answer is the first entry in the table, corresponding to serial history-based execution. In this mode, the cache works fine. However, we can see that even in serial, the event-based mode gives in correct results, as the logic is basically something like:

```C++
  for (Particle& p : particles) {
    p->event_calculate_xs();
  }
  ...

   for (Particle& p : particles) {
    p.event_collide();
  }
```

such that the thread will keep overwriting the MGXS cache for each particle in the first loop, and then erroneously apply the final state of the cache (generated when cross sections were looked up while processing the last particle in the first loop) to all particles in the second loop. Introducing multiple threads or processes causes even more issues.

I plan on fixing this issue with a PR shortly that will convert the per-thread MGXS cache to a particle-owned MGXS cache.

## Reproducer Code

```python
from math import log10

import numpy as np

import openmc
import openmc.mgxs

###############################################################################
# Create multigroup data

# Instantiate the energy group data
groups = openmc.mgxs.EnergyGroups(group_edges=[
    1e-5, 0.0635, 10.0, 1.0e2, 1.0e3, 0.5e6, 1.0e6, 20.0e6])

# Instantiate the 7-group (C5G7) cross section data
uo2_xsdata = openmc.XSdata('UO2', groups, temperatures=[294.0, 600.0])
uo2_xsdata.order = 0
scatter_matrix = np.array(
    [[[0.1275370, 0.0423780, 0.0000094, 0.0000000, 0.0000000, 0.0000000, 0.0000000],
      [0.0000000, 0.3244560, 0.0016314, 0.0000000, 0.0000000, 0.0000000, 0.0000000],
      [0.0000000, 0.0000000, 0.4509400, 0.0026792, 0.0000000, 0.0000000, 0.0000000],
      [0.0000000, 0.0000000, 0.0000000, 0.4525650, 0.0055664, 0.0000000, 0.0000000],
      [0.0000000, 0.0000000, 0.0000000, 0.0001253, 0.2714010, 0.0102550, 0.0000000],
      [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0012968, 0.2658020, 0.0168090],
      [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0085458, 0.2730800]]])
scatter_matrix = np.rollaxis(scatter_matrix, 0, 3)

# Original C5G7 data
uo2_xsdata.set_total([0.1779492, 0.3298048, 0.4803882, 0.5543674, 0.3118013, 0.3951678, 0.5644058], temperature=294.0)
uo2_xsdata.set_absorption([8.0248E-03, 3.7174E-03, 2.6769E-02, 9.6236E-02, 3.0020E-02, 1.1126E-01, 2.8278E-01], temperature=294.0)
uo2_xsdata.set_scatter_matrix(scatter_matrix, temperature=294.0)
uo2_xsdata.set_fission([7.21206E-03, 8.19301E-04, 6.45320E-03, 1.85648E-02, 1.78084E-02, 8.30348E-02, 2.16004E-01], temperature=294.0)
uo2_xsdata.set_nu_fission([2.005998E-02, 2.027303E-03, 1.570599E-02, 4.518301E-02, 4.334208E-02, 2.020901E-01, 5.257105E-01], temperature=294.0)
uo2_xsdata.set_chi([5.8791E-01, 4.1176E-01, 3.3906E-04, 1.1761E-07, 0.0000E+00, 0.0000E+00, 0.0000E+00], temperature=294.0)

# Altered C5G7 data (permuted Chi)
uo2_xsdata.set_total([0.1779492, 0.3298048, 0.4803882, 0.5543674, 0.3118013, 0.3951678, 0.5644058], temperature=600.0)
uo2_xsdata.set_absorption([8.0248E-03, 3.7174E-03, 2.6769E-02, 9.6236E-02, 3.0020E-02, 1.1126E-01, 2.8278E-01], temperature=600.0)
uo2_xsdata.set_scatter_matrix(scatter_matrix, temperature=600.0)
uo2_xsdata.set_fission([7.21206E-03, 8.19301E-04, 6.45320E-03, 1.85648E-02, 1.78084E-02, 8.30348E-02, 2.16004E-01], temperature=600.0)
uo2_xsdata.set_nu_fission([2.005998E-02, 2.027303E-03, 1.570599E-02, 4.518301E-02, 4.334208E-02, 2.020901E-01, 5.257105E-01], temperature=600.0)
uo2_xsdata.set_chi([4.1176E-01, 5.8791E-01, 3.3906E-04, 1.1761E-07, 0.0000E+00, 0.0000E+00, 0.0000E+00], temperature=600.0)


h2o_xsdata = openmc.XSdata('LWTR', groups)
h2o_xsdata.order = 0
h2o_xsdata.set_total([0.15920605, 0.412969593, 0.59030986, 0.58435,
                      0.718, 1.2544497, 2.650379])
h2o_xsdata.set_absorption([6.0105E-04, 1.5793E-05, 3.3716E-04,
                           1.9406E-03, 5.7416E-03, 1.5001E-02,
                           3.7239E-02])
scatter_matrix = np.array(
    [[[0.0444777, 0.1134000, 0.0007235, 0.0000037, 0.0000001, 0.0000000, 0.0000000],
      [0.0000000, 0.2823340, 0.1299400, 0.0006234, 0.0000480, 0.0000074, 0.0000010],
      [0.0000000, 0.0000000, 0.3452560, 0.2245700, 0.0169990, 0.0026443, 0.0005034],
      [0.0000000, 0.0000000, 0.0000000, 0.0910284, 0.4155100, 0.0637320, 0.0121390],
      [0.0000000, 0.0000000, 0.0000000, 0.0000714, 0.1391380, 0.5118200, 0.0612290],
      [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0022157, 0.6999130, 0.5373200],
      [0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.0000000, 0.1324400, 2.4807000]]])
scatter_matrix = np.rollaxis(scatter_matrix, 0, 3)
h2o_xsdata.set_scatter_matrix(scatter_matrix)

mg_cross_sections_file = openmc.MGXSLibrary(groups)
mg_cross_sections_file.add_xsdatas([uo2_xsdata, h2o_xsdata])
mg_cross_sections_file.export_to_hdf5()

###############################################################################
# Create materials for the problem

# Instantiate some Macroscopic Data
uo2_data = openmc.Macroscopic('UO2')
h2o_data = openmc.Macroscopic('LWTR')

# Instantiate some Materials and register the appropriate Macroscopic objects
uo2 = openmc.Material(name='UO2 fuel')
uo2.set_density('macro', 1.0)
uo2.add_macroscopic(uo2_data)

water = openmc.Material(name='Water')
water.set_density('macro', 1.0)
water.add_macroscopic(h2o_data)

# Instantiate a Materials collection and export to XML
materials_file = openmc.Materials([uo2, water])
materials_file.cross_sections = "mgxs.h5"
materials_file.export_to_xml()

###############################################################################
# Define problem geometry

# Create a surface for the fuel outer radius
fuel_ir = openmc.ZCylinder(r=0.25, name='Fuel IR')
fuel_or = openmc.ZCylinder(r=0.54, name='Fuel OR')

# Create a region represented as the inside of a rectangular prism
pitch = 1.26
box = openmc.rectangular_prism(pitch, pitch, boundary_type='reflective')

# Instantiate Cells
fuel_inner = openmc.Cell(fill=uo2, region=-fuel_ir, name='fuel inner')
fuel_inner.temperature = 600.0
fuel_outer = openmc.Cell(fill=uo2, region=+fuel_ir & -fuel_or, name='fuel outer')
fuel_outer.temperature = 294.0
moderator = openmc.Cell(fill=water, region=+fuel_or & box, name='moderator')

# Create a geometry with the two cells and export to XML
geometry = openmc.Geometry([fuel_inner, fuel_outer, moderator])
geometry.export_to_xml()

###############################################################################
# Define problem settings

# Instantiate a Settings object, set all runtime parameters, and export to XML
settings = openmc.Settings()
settings.energy_mode = "multi-group"
settings.batches = 10
settings.inactive = 5
settings.particles = 10000

# Create an initial uniform spatial source distribution over fissionable zones
lower_left = (-pitch/2, -pitch/2, -1)
upper_right = (pitch/2, pitch/2, 1)
uniform_dist = openmc.stats.Box(lower_left, upper_right, only_fissionable=True)
settings.source = openmc.IndependentSource(space=uniform_dist)
settings.export_to_xml()

###############################################################################
# Define tallies

# Create a mesh that will be used for tallying
mesh = openmc.RegularMesh()
mesh.dimension = (100, 100)
mesh.lower_left = (-pitch/2, -pitch/2)
mesh.upper_right = (pitch/2, pitch/2)

# Create a mesh filter that can be used in a tally
mesh_filter = openmc.MeshFilter(mesh)

# Now use the mesh filter in a tally and indicate what scores are desired
mesh_tally = openmc.Tally(name="Mesh tally")
mesh_tally.filters = [mesh_filter]
mesh_tally.scores = ['flux', 'fission', 'nu-fission']

# Let's also create a tally to get the flux energy spectrum. We start by
# creating an energy filter
e_min, e_max = 1e-5, 20.0e6
groups = 500
energies = np.logspace(log10(e_min), log10(e_max), groups + 1)
energy_filter = openmc.EnergyFilter(energies)

spectrum_tally = openmc.Tally(name="Flux spectrum")
spectrum_tally.filters = [energy_filter]
spectrum_tally.scores = ['flux']

# Instantiate a Tallies collection and export to XML
tallies = openmc.Tallies([mesh_tally, spectrum_tally])
tallies.export_to_xml()
```

### PR DESCRIPTION
**Pull request by jtramm**
# Overview

This PR converts per-thread multigroup cross section data caches into particle-owned caches, and is a partial remedy for #2160 and is a full fix for #2590. This PR closes #2590, but #2160 will still be left open as this PR does not solve all the performance issues.

As discussed in Issue #2160, the use of the per-thread cache involved access patterns like:

https://github.com/openmc-dev/openmc/blob/33ead74209d3290e7948110afc8b1d9fe84c1176/src/mgxs.cpp#L428-L439

This existing strategy was problematic for two reasons:

1) It is canonical multithreaded "false sharing" due to multiple cache objects fitting on the same cache line, but being read/written to by different threads (#2160).

2) It generates incorrect results in event-based mode (#2590), as per-thread particle states like this are meaningless.

To fix these issues, this PR simply moves the cache into the `ParticleData` class. I also experimented with completely removing the idea of caching MGXS data altogether, but found it was more optimal in terms of performance to retain the cache in the particle.

# Other Necessary Changes

## Number of Caches

The original scheme involved per-thread caches both for all materials and for all nuclides (if present). For simplicity, to reduce memory footprint, and because the determination of temperature and angle parameters is not that expensive, the new particle-owned scheme only maintains a single cache, which now corresponds to only the most recent material index. I.e., in the previous implementation, a particle might traverse from:

 Material A -> Material B -> Material A

without changing energy and be able to reuse a cache entry when re-entering Material A, as each thread had a separate cache for all materials. However, with the new scheme in this PR, the cache will be overwritten when the particle traverses from A -> B, such that determination of the temperature and angle indices will need to be redone when re-entering Material A. As we will see in the performance analysis section, the new scheme is overall faster, so it seems like it may be a net benefit to have more lookups if this saves memory space.

## Interfaces

Previously, the `MGXS::get_xs()` function that utilized caches had the following interface:

```C++
double Mgxs::get_xs(MgxsType xstype, int gin, const int* gout, const double* mu, const int* dg)
```

but this PR adds two integer arguments for the temperature index and angle index:

```C++
double Mgxs::get_xs(MgxsType xstype, int gin, const int* gout, const double* mu, const int* dg, int t, int a)
```

The means the caller needs to read/update the particle's cache themselves and provide this to the `MGXS::get_xs()` function. Thankfully, this is only done in two places:

1) as part of the `Mgxs::calculate_xs(Particle& p)` function
2) when scoring tallies in the `score_general_mg()` function

While refactoring of the `calculate_xs()` function was straightforward, the refactoring of the tally scoring routines ended up being more painful, as each score typically makes a few calls to `get_xs()`. 

# Correctness

I re-used the multi-group multi-temperature pincell from #2590 and ran the same tests on the new branch. As shown in the table below, the parallel event-based mode now gives the correct answer (i.e., the same as if running with a serial thread in history-based mode). Therefore -- this PR closes #2590.

Parallelism | Branch | Mode | k-eff
-- | -- | -- | --
Serial | Thread Cache (Baseline) | history | 1.32339
Serial | Thread Cache (Baseline) | event | 1.32117
MPI 112 Ranks | Thread Cache (Baseline) | event | 1.31853
OpenMP 112 Threads | Thread Cache (Baseline) | event | 1.32285
OpenMP 112 Threads | Particle Cache (This PR) | event | 1.32339
OpenMP 112 Threads | No Cache | event | 1.32339

# Performance Analysis

I tested performance of the PR against develop on the C5G7 benchmark. I also tested a version of the code that does not have any MGXS cache at all (i.e., the angle and temperature indices are looked up on-the-fly each time they are needed). Given that the C5G7 MGXS data is isotropic and isothermal, this was expected to be fast, but it appeared to have a performance cost to it.

 I ran each configuration 10 times to get a sense of the variance in performance timings. I tested using two test systems: a dual-socket Xeon 8180M with 56 cores/112 threads total, and a dual-socket Xeon 6336Y node with 48 cores and 96 threads total. The 8180M node has been the only node so far that has shown the poor thread scaling performance, whereas the 6336Y node behaves well (as most architectures I've tested do on this problem).

2D C5G7 - 5 inactive - 5 active - 100k particles/batch - gcc 12.2.0 - 2x 8180M (56c/112t)

Parallelism | Branch | Inactive | std. dev. | Active | std. dev.
-- | -- | -- | -- | -- | --
Serial | Thread Cache (Baseline) | 39,904 | 104 | 36,669 | 97
MPI - 112 Ranks | Thread Cache (Baseline) | 1,351,320 | 266,381 | 957,790 | 119,047
OpenMP - 112 Threads | Thread Cache (Baseline) | 166,462 | 8,084 | 95,067 | 2,244
OpenMP - 112 Threads | Particle Cache (This PR) | 176,075 | 11,243 | 99,457 | 3,963
OpenMP - 112 Threads | No Cache | 152,166 | 16,650 | 96,740 | 4,366

2D C5G7 - 5 inactive - 5 active - 100k particles/batch - gcc 12.2.0 - 2x 6336Y (48c/96t)

Parallelism | Branch | Inactive | std. dev. | Active | std. dev.
-- | -- | -- | -- | -- | --
Serial | Thread Cache (Baseline) | 44,123 | 164 | 40,334 | 136
MPI - 112 Ranks | Thread Cache (Baseline) | 1,445,500 | 47,892 | 1,028,990 | 12,994
OpenMP - 112 Threads | Thread Cache (Baseline) | 1,084,334 | 22,947 | 762,377 | 8,591
OpenMP - 112 Threads | Particle Cache (This PR) | 1,008,600 | 7,293 | 742,361 | 9,525
OpenMP - 112 Threads | No Cache | 854,124 | 28,345 | 656,255 | 21,873

Basically, conversion of the thread-owned XS cache to a particle-owned one does not really improve the performance picture. Moving the cache to the particle has a minor speedup on the troublesome 8180M system, but also shows a minor slowdown on the 6336Y node. Removing the cache completely was also not a great idea, as this performed much slower than the particle-owned cache on both systems. Thus,  #2160 will need to remain open as this PR did not really help with whatever the underlying cause of poor thread scaling on the 8180M is.

# Other Potential Areas of Memory Contention to Investigate on the 8180M node

There are a few other areas that may have been causing memory contention:

1) The neighbor list (due to locking when appending to a cell's neighbor list)
2) The shared fission bank (due to atomically incrementing the size integer when appending). This is relatively much more costly given how much faster MG mode usually is compared to CE, so contention over this variable is more of an issue.
3) `Particle` initialization (due to calls to `New` or `malloc` in its associated constructors)

To test out if these were the root causes, I implemented algorithms to remove all contention from these areas, but did not observe any additional speedups. Analysis with `perf` and `vtune` was somewhat confounding, as they may not have been running correctly on this system due to cluster permissions issues so were identifying seemingly odd locations as being impractically expensive. Manual addition of history-based mode OpenMP timers (and examining performance timers in event-based mode) revealed that basically all the major areas of the code were running slow when a lot of threads were running. I.e., the slowdown did not appear to be limited to a particular location, rather, the entire program seemed to be running slowly.

I also experimented with using `numactl` to manually manage where the allocations were happening and the thread affinity policies, and while these had marginal benefit, but did not explain the overall poor thread scaling. 

Overall, my guess is that there may be a system configuration issue with the 8180M node that is affecting us for some reason, or it may simply be due to a quirk of the cache architecture of skylake that has been remedied in newer architectures since then (as 2017's 8180M is now 6 years old). As discussed in #2160, other dual-socket CPUs I've tested on have had good thread scaling, even before this PR, so right now it doesn't seem worth worrying about further. However, if we see similar thread scaling issues on newer chips, I will plan to commit more time resources to solving this issue. For now, I will consider this a partial remedy for #2160, so will leave that issue open.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: include/openmc/mgxs.h [hunk 1]
```diff
@@ -159,23 +149,37 @@ class Mgxs {
...
+  //!
+  //! @param p Particle.
+  void set_temperature_index(Particle& p);
+
+  //! \brief Gets the temperature index given a temperature.
   //!
   //! @param sqrtkT Temperature of the material.
-  void set_temperature_index(double sqrtkT);
+  //! @return The temperature index corresponding to sqrtkT.
+  int get_temperature_index(double sqrtkT);
```
Comment (paulromano) [inline include/openmc/mgxs.h:171]:
```suggestion
  int get_temperature_index(double sqrtkT) const;
```

File: include/openmc/mgxs.h [hunk 2]
```diff
@@ -159,23 +149,37 @@ class Mgxs {
...
+  //! @param p Particle.
+  void set_angle_index(Particle& p);
 
-  //! \brief Sets the angle index in cache given a direction
+  //! \brief Gets the angle index given a direction.
   //!
   //! @param u Incoming particle direction.
-  void set_angle_index(Direction u);
+  //! @return The angle index corresponding to u.
+  int get_angle_index(Direction& u);
```
Comment (paulromano) [inline include/openmc/mgxs.h:182]:
```suggestion
  int get_angle_index(const Direction& u) const;
```

File: src/mgxs.cpp [hunk 3]
```diff
@@ -643,32 +621,26 @@ bool Mgxs::equiv(const Mgxs& that)
 
 //==============================================================================
 
-void Mgxs::set_temperature_index(double sqrtkT)
+int Mgxs::get_temperature_index(double sqrtkT)
```
Comment (paulromano) [inline src/mgxs.cpp:624]:
```suggestion
int Mgxs::get_temperature_index(double sqrtkT) const
```

File: src/mgxs.cpp [hunk 4]
```diff
@@ -643,32 +621,26 @@ bool Mgxs::equiv(const Mgxs& that)
...
-#endif
-  if (!is_isotropic && ((u.x != cache[tid].u) || (u.y != cache[tid].v) ||
-                         (u.z != cache[tid].w))) {
+  p.mg_xs_cache().t = get_temperature_index(p.sqrtkT());
+  p.mg_xs_cache().sqrtkT = p.sqrtkT();
+}
+
+//==============================================================================
+
+int Mgxs::get_angle_index(Direction& u)
```
Comment (paulromano) [inline src/mgxs.cpp:639]:
```suggestion
int Mgxs::get_angle_index(const Direction& u) const
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for this fix @jtramm and the very thorough explanation of the whole situation! Other than the very minor line comments below, one thing that I would like to see before we merge is the addition of a test to cover this case so we don't mess it up in the future.

**Comment by jtramm** [comment]
I've added a test in, and can confirm that the test fails with develop in event-based mode (e.g., `pytest --event`) but passes with this branch in both history- and event-based modes.

The CI is giving failures that seem to be unrelated to anything I've added:
```
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for h5py
ERROR: Could not build wheels for h5py, which is required to install pyproject.toml-based projects
Failed to build h5py

Notice:  A new release of pip is available: 23.1.2 -> 23.2
Notice:  To update, run: pip install --upgrade pip
Error: Process completed with exit code 1.
```

Any ideas on how to resolve the CI issue?

**Comment by paulromano** [comment]
Argh, sometimes I hate CI :disappointed: I'll look into that and see if I can get to the bottom of it

**Comment by paulromano** [comment]
@jtramm I think you'll need to push a new commit to this branch to get CI passing (can even be an empty commit)

**Comment by jtramm** [comment]
I added in a comment-only commit -- looks like CI is passing now.

**Review by paulromano — APPROVED** [review]
Thanks @jtramm!

**Pull request merged**